### ISSUE TITLE
Segfault with intervention/image test suite

### ISSUE THREAD

**Issue by remicollet** [mkoppanen/imagick#195]
From #183 

```
(gdb) bt
#0  zend_mm_alloc_small (bin_num=16, size=<optimized out>, heap=0x7f4a7dc00040) at /usr/src/debug/php-7.0.15/Zend/zend_alloc.c:1306
#1  zend_mm_alloc_heap (size=<optimized out>, heap=0x7f4a7dc00040) at /usr/src/debug/php-7.0.15/Zend/zend_alloc.c:1377
#2  _emalloc (size=<optimized out>) at /usr/src/debug/php-7.0.15/Zend/zend_alloc.c:2461
#3  0x000056191db7ffe7 in zend_hash_real_init_ex (packed=0, ht=0x7f4a73899f50) at /usr/src/debug/php-7.0.15/Zend/zend_hash.c:140
#4  zend_hash_check_init (packed=0, ht=0x7f4a73899f50) at /usr/src/debug/php-7.0.15/Zend/zend_hash.c:163
#5  _zend_hash_add_or_update_i (flag=2, pData=0x7f4a7dc171b0, key=0x7f4a7dcb35c0, ht=0x7f4a73899f50) at /usr/src/debug/php-7.0.15/Zend/zend_hash.c:563
#6  _zend_hash_add (ht=ht@entry=0x7f4a73899f50, key=0x7f4a7dcb35c0, pData=pData@entry=0x7f4a7dc171b0) at /usr/src/debug/php-7.0.15/Zend/zend_hash.c:638
#7  0x000056191db7198f in zval_copy_static_var (p=0x7f4a7dc171b0, num_args=<optimized out>, args=<optimized out>, key=0x7ffdc12ba530)
    at /usr/src/debug/php-7.0.15/Zend/zend_variables.c:317
#8  0x000056191db85769 in zend_hash_apply_with_arguments (ht=ht@entry=0x7f4a7dd8faf0, apply_func=0x56191db71930 <zval_copy_static_var>, num_args=num_args@entry=1)
    at /usr/src/debug/php-7.0.15/Zend/zend_hash.c:1597
#9  0x000056191db9d909 in zend_create_closure (res=<optimized out>, func=0x7f4a73846b60, scope=0x7f4a73844918, called_scope=0x7f4a73844918, this_ptr=this_ptr@entry=0x0)
    at /usr/src/debug/php-7.0.15/Zend/zend_closures.c:573
#10 0x000056191dbc30ca in ZEND_DECLARE_LAMBDA_FUNCTION_SPEC_CONST_UNUSED_HANDLER () at /usr/src/debug/php-7.0.15/Zend/zend_vm_execute.h:8189
#11 0x000056191dbb3b7b in execute_ex (ex=<optimized out>) at /usr/src/debug/php-7.0.15/Zend/zend_vm_execute.h:414
#12 0x000056191db64458 in zend_call_function (fci=fci@entry=0x7ffdc12ba840, fci_cache=0x7f4a7dc17130, fci_cache@entry=0x7ffdc12ba810)
    at /usr/src/debug/php-7.0.15/Zend/zend_execute_API.c:858
#13 0x000056191da61729 in zim_reflection_method_invokeArgs (execute_data=<optimized out>, return_value=0x7f4a7dc16700)
    at /usr/src/debug/php-7.0.15/ext/reflection/php_reflection.c:3355
#14 0x000056191dbf91ee in ZEND_DO_FCALL_SPEC_HANDLER () at /usr/src/debug/php-7.0.15/Zend/zend_vm_execute.h:842
#15 0x000056191dbb3b7b in execute_ex (ex=<optimized out>) at /usr/src/debug/php-7.0.15/Zend/zend_vm_execute.h:414
#16 0x000056191dc08287 in zend_execute (op_array=op_array@entry=0x7f4a7dc86000, return_value=return_value@entry=0x7f4a7dd99400)
    at /usr/src/debug/php-7.0.15/Zend/zend_vm_execute.h:458
#17 0x000056191db73313 in zend_execute_scripts (type=2109826368, type@entry=8, retval=0x7f4a7dd99400, retval@entry=0x0, file_count=file_count@entry=3)
    at /usr/src/debug/php-7.0.15/Zend/zend.c:1437
#18 0x000056191db127b0 in php_execute_script (primary_file=0x7ffdc12bcf00) at /usr/src/debug/php-7.0.15/main/main.c:2492
#19 0x000056191dc09f3c in do_cli (argc=11, argv=0x56191e88e1e0) at /usr/src/debug/php-7.0.15/sapi/cli/php_cli.c:977
#20 0x000056191d9f2569 in main (argc=11, argv=0x56191e88e1e0) at /usr/src/debug/php-7.0.15/sapi/cli/php_cli.c:1347
```

Using: 
```
PHP Version => 7.0.15

imagick module => enabled
imagick module version => 3.4.3RC3
imagick classes => Imagick, ImagickDraw, ImagickPixel, ImagickPixelIterator, ImagickKernel
Imagick compiled with ImageMagick version => ImageMagick 6.9.7-5 Q16 x86_64 2017-01-22 http://www.imagemagick.org
Imagick using ImageMagick library version => ImageMagick 6.9.7-5 Q16 x86_64 2017-01-22 http://www.imagemagick.org

```

**Comment by remicollet** [comment]
Notice: this only happens with PHP 7, no segfault with PHP 5

**Comment by Danack** [comment]
Thanks. That's quite odd. 

I can't think of anything in Imagick or PHP that has changed recently that would cause this, and yet there it is.

Obviously, it's also not a simple Imagick thing that's failing....it's an internal PHP thing. Which are always fun to debug...

**Comment by remicollet** [comment]
@Danack see the PR #196   ;)    (thanks to valgrind)

**Comment by Danack** [comment]
You're too quick - was still reading and trying to understand it, so didn't see your PR.

Presumably, this issue isn't something that broken recently then. It's just always been broken on PHP 7 and this is the first time that someone has noticed?

**Comment by remicollet** [comment]
1/ intervention/image team is aware of this issue for a long time => https://travis-ci.org/Intervention/image

and they sadly never think to open a issue here :(

2/ test suite and class mocking are horrible stuff which can raise strange issue (especially creating new object without calling the constructor. Not real usage.

**Comment by Danack** [comment]
> 1/ intervention/image team is aware of this issue for a long time 

https://www.youtube.com/watch?v=5vrXKlO2Jbw

Even if the use case is horrible, I'm still going to attempt to write a test case for this.

I'll merge the PR and do another RC when I get back from work.

**Comment by Danack** [comment]
I couldn't make an exact reproduce case, but below is similar.

```
<?php

interface MockInterface {}

class ImagickPixelMock extends \ImagickPixel implements MockInterface
{
    protected $foo1;
    protected $foo2;
    protected $foo3;
    protected $foo4;

    public function __isset($name)
    {
    }
}

$reflectionClass = new ReflectionClass('ImagickPixelMock');
$instance = $reflectionClass->newInstanceWithoutConstructor();

$fn = function () use ($instance) {
    $method = new ReflectionMethod($instance, 'getcolorvalue');
    $method->invokeArgs($instance, [\Imagick::COLOR_GREEN]);
};

$fn();

```

**Comment by remicollet** [comment]
This reporoducer segfaults but differently (with ot without the patch)

> php: wand/pixel-wand.c:999: PixelGetGreen: Assertion `wand != (const PixelWand *) NULL' failed.

(gdb) bt
#0  0x00007ffff4a8991f in raise () at /lib64/libc.so.6
#1  0x00007ffff4a8b51a in abort () at /lib64/libc.so.6
#2  0x00007ffff4a81da7 in __assert_fail_base () at /lib64/libc.so.6
#3  0x00007ffff4a81e52 in  () at /lib64/libc.so.6
#4  0x00007fffce8388af in PixelGetGreen (wand=<optimized out>) at wand/pixel-wand.c:999
#5  0x00007fffceaae8d5 in zim_imagickpixel_getcolorvalue (execute_data=0x7ffff38132f0, return_value=0x7fffffff9e50)
    at /usr/src/debug/php-pecl-imagick-3.4.3/NTS/imagickpixel_class.c:427
#6  0x00007fffb472f498 in pt_execute_core (internal=1, execute_data=0x7ffff38132f0, return_value=0x7fffffff9e50)
    at /usr/src/debug/php-pecl-trace-1.0.0/trace-1.0.0/extension/trace.c:1316
#7  0x0000555555791662 in zend_call_function (fci=fci@entry=0x7fffffff9e90, fci_cache=fci_cache@entry=0x7fffffff9e60) at /usr/src/debug/php-7.0.15/Zend/zend_execute_API.c:880
#8  0x000055555568e729 in zim_reflection_method_invokeArgs (execute_data=<optimized out>, return_value=0x7ffff3813260)
    at /usr/src/debug/php-7.0.15/ext/reflection/php_reflection.c:3355
#9  0x00007fffb472f498 in pt_execute_core (internal=1, execute_data=0x7ffff3813270, return_value=0x7ffff3813260)
    at /usr/src/debug/php-pecl-trace-1.0.0/trace-1.0.0/extension/trace.c:1316
#10 0x0000555555825c70 in ZEND_DO_FCALL_SPEC_HANDLER () at /usr/src/debug/php-7.0.15/Zend/zend_vm_execute.h:844
#11 0x00005555557e0b7b in execute_ex (ex=<optimized out>) at /usr/src/debug/php-7.0.15/Zend/zend_vm_execute.h:414
#12 0x00007fffb472f66b in pt_execute_core (internal=0, execute_data=0x7ffff3813170, return_value=0x0) at /usr/src/debug/php-pecl-trace-1.0.0/trace-1.0.0/extension/trace.c:1321
#13 0x0000555555825dad in ZEND_DO_FCALL_SPEC_HANDLER () at /usr/src/debug/php-7.0.15/Zend/zend_vm_execute.h:800
#14 0x00005555557e0b7b in execute_ex (ex=<optimized out>) at /usr/src/debug/php-7.0.15/Zend/zend_vm_execute.h:414
#15 0x00007fffb472f66b in pt_execute_core (internal=0, execute_data=0x7ffff3813030, return_value=0x0) at /usr/src/debug/php-pecl-trace-1.0.0/trace-1.0.0/extension/trace.c:1321
#16 0x0000555555835287 in zend_execute (op_array=op_array@entry=0x7ffff3872000, return_value=return_value@entry=0x7ffff38df340)
    at /usr/src/debug/php-7.0.15/Zend/zend_vm_execute.h:458
#17 0x00005555557a0313 in zend_execute_scripts (type=-209637008, type@entry=8, retval=0x7ffff38df340, retval@entry=0x0, file_count=file_count@entry=3)
    at /usr/src/debug/php-7.0.15/Zend/zend.c:1437
#18 0x000055555573f7b0 in php_execute_script (primary_file=0x7fffffffca70) at /usr/src/debug/php-7.0.15/main/main.c:2492
#19 0x0000555555836f3c in do_cli (argc=2, argv=0x555555ba7200) at /usr/src/debug/php-7.0.15/sapi/cli/php_cli.c:977
#20 0x000055555561f569 in main (argc=2, argv=0x555555ba7200) at /usr/src/debug/php-7.0.15/sapi/cli/php_cli.c:1347

**Comment by remicollet** [comment]
Perhaps we need something like 
https://github.com/php/php-src/blob/master/ext/date/php_date.c#L628
```

#define DATE_CHECK_INITIALIZED(member, class_name) \
	if (!(member)) { \
		php_error_docref(NULL, E_WARNING, "The " #class_name " object has not been correctly initialized by its constructor"); \
		RETURN_FALSE; \
	}

```

**Comment by Danack** [comment]
The example I posted seems to be critically dependent on the number of properties/methods in the object.

If I alter the number of properties/methods I sometimes see the exception stack you posted which reaches `PixelGetGreen`, but other times I see the stack trace I posted which falls over before then.


> Perhaps we need something like
> https://github.com/php/php-src/blob/master/ext/date/php_date.c#L628

Yeah. The Imagick class already has that with `php_imagick_ensure_not_empty()` - I'm not sure why the ImagickPixel or other classes don't also have it.

I'll probably hold off doing that sanity check until the next release, as touching so many lines when people are waiting for a release, and we're in an RC phase seems...sub-optimal.
